pilot/pkg/xds/sds.go

```diff
@@ -102,50 +103,101 @@ func (s *SecretGen) proxyAuthorizedForSecret(proxy *model.Proxy, sr SecretResour
+	if proxy.MergedGateway == nil {
+		return CrossNamespaceCertificateReferences
+	}
+	verifiedCertificateReferences := proxy.MergedGateway.VerifiedCertificateReferences
+	for _, r := range resources {
+		if !verifiedCertificateReferences.Contains(r.Name) {
+			return CrossNamespaceCertificateReferences
+		}
+	}
+	return OnlyLocalCertificateReferences
+}
+
 func (s *SecretGen) Generate(proxy *model.Proxy, push *model.PushContext, w *model.WatchedResource,
 	req *model.PushRequest) (model.Resources, model.XdsLogDetails, error) {
 	if proxy.VerifiedIdentity == nil {
 		log.Warnf("proxy %v is not authorized to receive secrets. Ensure you are connecting over TLS port and are authenticated.", proxy.ID)
 		return nil, model.DefaultXdsLogDetails, nil
 	}
-	secrets, err := s.secrets.ForCluster(proxy.Metadata.ClusterID)
-	if err != nil {
-		log.Warnf("proxy %v is from an unknown cluster, cannot retrieve certificates: %v", proxy.ID, err)
-		return nil, model.DefaultXdsLogDetails, nil
-	}
-	if err := secrets.Authorize(proxy.VerifiedIdentity.ServiceAccount, proxy.VerifiedIdentity.Namespace); err != nil {
-		log.Warnf("proxy %v is not authorized to receive secrets: %v", proxy.ID, err)
-		return nil, model.DefaultXdsLogDetails, nil
-	}
 	if req == nil || !needsUpdate(proxy, req.ConfigsUpdated) {
 		return nil, model.DefaultXdsLogDetails, nil
 	}
 	var updatedSecrets map[model.ConfigKey]struct{}
 	if !req.Full {
 		updatedSecrets = model.ConfigsOfKind(req.ConfigsUpdated, gvk.Secret)
 	}
+
+	// TODO: For the new gateway-api, we should always search the config namespace and stop reading across all clusters
+	secrets, err := s.secrets.ForCluster(proxy.Metadata.ClusterID)
+	if err != nil {
+		log.Warnf("proxy %v is from an unknown cluster, cannot retrieve certificates: %v", proxy.ID, err)
+		return nil, model.DefaultXdsLogDetails, nil
+	}
+
+	resources := parseResources(w.ResourceNames, proxy)
+	authMode := getAuthorizationMode(resources, proxy)
+	if authMode != OnlyLocalCertificateReferences {
+		if err := secrets.Authorize(proxy.VerifiedIdentity.ServiceAccount, proxy.VerifiedIdentity.Namespace); err != nil {
+			log.Warnf("proxy %v is not authorized to receive secrets: %v", proxy.ID, err)
+			return nil, model.DefaultXdsLogDetails, nil
+		}
+	}
+
 	results := model.Resources{}
 	cached, regenerated := 0, 0
-	for _, resource := range w.ResourceNames {
-		sr, err := parseResourceName(resource, proxy.ConfigNamespace, string(proxy.Metadata.ClusterID))
-		if err != nil {
-			pilotSDSCertificateErrors.Increment()
-			log.Warnf("error parsing resource name: %v", err)
-			continue
-		}
 
+	for _, sr := range resources {
 		if updatedSecrets != nil {
 			if !containsAny(updatedSecrets, relatedConfigs(model.ConfigKey{Kind: gvk.Secret, Name: sr.Name, Namespace: sr.Namespace})) {
 				// This is an incremental update, filter out secrets that are not updated.
 				continue
 			}
 		}
 
-		if err := s.proxyAuthorizedForSecret(proxy, sr); err != nil {
```

Comment:
Shouldn't we still do this for the `CrossNamespaceCertificateReferences` case?

Reply:
We should - and we still do (we have tests as well). This however is dead code, so I removed the function. The new check is where we have `requested secret %v not accessible for proxy %v: SDS is currently only supporting accessing secret within the same namespace. `.

Would be good to have a sanity check that actually makes sense - coming back to this 2 weeks later I think it is but this is critical code